### DISCUSSION TITLE
Idea: virtual spaces

### DISCUSSION THREAD

**Discussion by AnatolyRugalev** [asmvik/yabai#1588] [Ideas]
I have an idea in mind how proper space management can be achieved in MacOS, but I lack skills to implement / test it efficiently. Currently Yabai needs to SiP protection to be disabled, and this, unfortunately, not always possible, plus, MacOS spaces design is frustrating.

What if we ignore MacOS spaces alltogether, and all window management happens inside a single space (or, one space per display). Can we create "virtual" spaces, so we can manipulate and move them between screens in no time, by just moving all virtual space windows individually? So, technically, all windows will be piled up onto each other in one MacOS space. To fix an issue with windows from other spaces being visible, we can also have special full-screen window which will create fake background between spaces. One such window per screen would be required. Here's a small illustration of that:
![image](https://user-images.githubusercontent.com/1397674/210380463-eceeb185-4e06-4142-9a79-cc37b1684abe.png)

In this concept one MacOS space is always kept on one screen, essentially, representing a screen itself. One thing that could be annoying is if you accidentally switch MacOS space on a screen to something else, ending up in a "nowhere" space. I can imagine that such idea has been considered previously, but couldn't find any evidence of that. My guess is that moving a bunch of windows individually could create some visual disturbance / glitches due to this method being slow.

Any thoughts on this?

**Comment by asmvik** [comment]
Mentioned before in https://github.com/koekeishiya/yabai/issues/1039 and https://github.com/koekeishiya/yabai/discussions/1414

The short answer is: Yes it is technically possible, but it's just not in the scope of what I want yabai to be.
